### DISCUSSION TITLE
Simulation from midway

### DISCUSSION THREAD

**Discussion by ytimber** [InteractiveComputerGraphics/SPlisHSPlasH#145] [Q&A] [answered]
Hi,

I'm wondering whether SPlisHSPlasH can realize the following request: 
Could I pass in parameters as a scene configuration file from a previous frame and continue simulation from this previous frame, rather than initializing and simulating everything from the very beginning? Many thanks in advance. It would mean a lot to me.

**Comment by janbender** [comment, marked as answer]
At any time in the simulation you can save the current state by pressing the key "s". You can load the state in Windows by pressing "l". Or you can load the state with the command line parameter "–state-file". A state file stores all parameter settings and the current state of all particles.

See also here: https://splishsplash.readthedocs.io/en/latest/getting_started.html

**Reply by ytimber** [reply]
@janbender 
Thank you so much for the quick reply.
I’m wondering instead of using hot key to save the specific state, could I automatically save a state that I like in off screen simulation? Thank you!

**Comment by janbender** [comment]
You can also export a state each X frames. Take a look at the scene file option "enableStateExport": 

https://splishsplash.readthedocs.io/en/latest/file_format.html#export

**Reply by ytimber** [reply]
@janbender 
Thank you so much for your quick reply! It works!! :))

**Reply by ytimber** [reply]
@janbender 
Another really quick question. I had a hard time wrapping my mind around how the state file is named. I find it a little arbitrary. It would mean a lot if you could elaborate a little. Many thanks!

**Reply by ytimber** [reply]
@janbender 
Sorry to bother you with another quick question. Can I load in the state file and update viscosity/other parameters for continuing simulation?

**Reply by janbender** [reply]
> @janbender Another really quick question. I had a hard time wrapping my mind around how the state file is named. I find it a little arbitrary. It would mean a lot if you could elaborate a little. Many thanks!

The name is simply: state_<Simulation time of the current state>.bin

**Reply by janbender** [reply]
> @janbender Sorry to bother you with another quick question. Can I load in the state file and update viscosity/other parameters for continuing simulation?

The simple solution is to pause the simulation when saving the state. Then you can load it, change the parameters and continue. 

You can also change a parameter in the command line (–param arg), see hjere:
https://splishsplash.readthedocs.io/en/latest/getting_started.html#command-line-options

However, I didn't test this for a long time and I changed many things in the simulator. So I'm not sure if currently the command line parameter overwrites the state file or the state file the parameter. Please give me some feedback if you test this option.

**Reply by ytimber** [reply]
@janbender 
Thank you for your reply.
For the state file, I didn't put it explicit enough, I meant the number string in the name after ```state_```. What do those numbers represent?
I will test the overwrite and let you know~!

**Comment by janbender** [comment]
If you save the state at simulation time 1.05s, you will get: state_1.05.bin

**Reply by ytimber** [reply]
@janbender 
Just to double check if I use fps=25 and save 25 times per second, then the name should be the time in seconds when the frame was saved. Therefore, I'm wondering whether there is any possible and easy way for me (such as not involving writing a new state exporter) to change the name of the exported state file to something like names saved by rigid body export (e.g. name with frame number)? Many thanks!!!

**Reply by ytimber** [reply]
@janbender 
I just tested the overwrite, the behavior of fluid still follows viscosity=0.01, which is the original viscosity set in scene configuration file, after I passed in the parameter ```--param Fluid:viscosity:100```

**Reply by janbender** [reply]
> @janbender Just to double check if I use fps=25 and save 25 times per second, then the name should be the time in seconds when the frame was saved. Therefore, I'm wondering whether there is any possible and easy way for me (such as not involving writing a new state exporter) to change the name of the exported state file to something like names saved by rigid body export (e.g. name with frame number)? Many thanks!!!

You can simply change the file name in this line of the code: 
https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/blob/95760d95c8084590a96ed1e91e9bef7a56c1297e/Simulator/SimulatorBase.cpp#L1638

**Reply by janbender** [reply]
> @janbender I just tested the overwrite, the behavior of fluid still follows viscosity=0.01, which is the original viscosity set in scene configuration file, after I passed in the parameter `--param Fluid:viscosity:100`

Okay, I will have a look in the next days.

**Reply by ytimber** [reply]
@janbender 
Thanks for pointing out the place for changing the code and many thanks for looking into updating viscosity midway. It means a lot for me!

**Reply by janbender** [reply]
> @janbender I just tested the overwrite, the behavior of fluid still follows viscosity=0.01, which is the original viscosity set in scene configuration file, after I passed in the parameter `--param Fluid:viscosity:100`

I fixed that yesterday. So now it should work.

**Reply by ytimber** [reply]
@janbender 
Thank you so much! I just tested -- it works!

**Reply by ytimber** [reply]
@janbender 
Hi, I just encountered this problem that I have no idea how to solve. 
I used the following command to do an offscreen simulation based on a past state, but with viscosity overwritten, 
```
MESA_GL_VERSION_OVERRIDE=3.3 ./SPHSimulator /home/ccn/Downloads/SPlisHSPlasH/data/Scenes/box10.json --no-gui --no-initial-pause --stopAt 1 --state-file /home/ccn/Downloads/SPlisHSPlasH/bin/output/box10/state/state_0.600492.bin
```
However, it returns with an error message: 
```Error: StopAt parameter must be set when starting without GUI.```
Didn't ```--stopAt 1 ``` set the stop at parameter? This problem was not encountered when I wasn't using ```state-file``` argument in the command line. Many thanks for your help in advance! :)

**Reply by janbender** [reply]
Now this should be fixed.

**Reply by ytimber** [reply]
@janbender 
Thank you so much! :))
